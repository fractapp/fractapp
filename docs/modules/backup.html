<!doctype html>
<html class="default no-js">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<title>Backup | fractapp</title>
	<meta name="description" content="Documentation for fractapp">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="stylesheet" href="../assets/css/main.css">
</head>
<body>
<header>
	<div class="tsd-page-toolbar">
		<div class="container">
			<div class="table-wrap">
				<div class="table-cell" id="tsd-search" data-index="../assets/js/search.json" data-base="..">
					<div class="field">
						<label for="tsd-search-field" class="tsd-widget search no-caption">Search</label>
						<input id="tsd-search-field" type="text" />
					</div>
					<ul class="results">
						<li class="state loading">Preparing search index...</li>
						<li class="state failure">The search index is not available</li>
					</ul>
					<a href="../index.html" class="title">fractapp</a>
				</div>
				<div class="table-cell" id="tsd-widgets">
					<div id="tsd-filter">
						<a href="#" class="tsd-widget options no-caption" data-toggle="options">Options</a>
						<div class="tsd-filter-group">
							<div class="tsd-select" id="tsd-filter-visibility">
								<span class="tsd-select-label">All</span>
								<ul class="tsd-select-list">
									<li data-value="public">Public</li>
									<li data-value="protected">Public/Protected</li>
									<li data-value="private" class="selected">All</li>
								</ul>
							</div>
							<input type="checkbox" id="tsd-filter-inherited" checked />
							<label class="tsd-widget" for="tsd-filter-inherited">Inherited</label>
							<input type="checkbox" id="tsd-filter-externals" checked />
							<label class="tsd-widget" for="tsd-filter-externals">Externals</label>
							<input type="checkbox" id="tsd-filter-only-exported" />
							<label class="tsd-widget" for="tsd-filter-only-exported">Only exported</label>
						</div>
					</div>
					<a href="#" class="tsd-widget menu no-caption" data-toggle="menu">Menu</a>
				</div>
			</div>
		</div>
	</div>
	<div class="tsd-page-title">
		<div class="container">
			<ul class="tsd-breadcrumb">
				<li>
					<a href="../globals.html">Globals</a>
				</li>
				<li>
					<a href="backup.html">Backup</a>
				</li>
			</ul>
			<h1>Namespace Backup</h1>
		</div>
	</div>
</header>
<div class="container container-main">
	<div class="row">
		<div class="col-8 col-content">
			<section class="tsd-panel tsd-comment">
				<div class="tsd-comment tsd-typography">
					<dl class="tsd-comment-tags">
						<dt>namespace</dt>
						<dd></dd>
					</dl>
				</div>
			</section>
			<section class="tsd-panel-group tsd-index-group">
				<h2>Index</h2>
				<section class="tsd-panel tsd-index-panel">
					<div class="tsd-index-content">
						<section class="tsd-index-section ">
							<h3>Enumerations</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-enum tsd-parent-kind-namespace"><a href="../enums/backup.backuptype.html" class="tsd-kind-icon">Backup<wbr>Type</a></li>
							</ul>
						</section>
						<section class="tsd-index-section ">
							<h3>Variables</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-variable tsd-parent-kind-namespace"><a href="backup.html#googledrivefolder" class="tsd-kind-icon">Google<wbr>Drive<wbr>Folder</a></li>
								<li class="tsd-kind-variable tsd-parent-kind-namespace"><a href="backup.html#algorithm" class="tsd-kind-icon">algorithm</a></li>
							</ul>
						</section>
						<section class="tsd-index-section ">
							<h3>Functions</h3>
							<ul class="tsd-index-list">
								<li class="tsd-kind-function tsd-parent-kind-namespace"><a href="backup.html#backup-1" class="tsd-kind-icon">backup</a></li>
								<li class="tsd-kind-function tsd-parent-kind-namespace"><a href="backup.html#decrypt" class="tsd-kind-icon">decrypt</a></li>
								<li class="tsd-kind-function tsd-parent-kind-namespace"><a href="backup.html#encrypt" class="tsd-kind-icon">encrypt</a></li>
								<li class="tsd-kind-function tsd-parent-kind-namespace"><a href="backup.html#getfile" class="tsd-kind-icon">get<wbr>File</a></li>
								<li class="tsd-kind-function tsd-parent-kind-namespace"><a href="backup.html#getseed" class="tsd-kind-icon">get<wbr>Seed</a></li>
							</ul>
						</section>
					</div>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Variables</h2>
				<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-namespace">
					<a name="googledrivefolder" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagConst">Const</span> Google<wbr>Drive<wbr>Folder</h3>
					<div class="tsd-signature tsd-kind-icon">Google<wbr>Drive<wbr>Folder<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">"fractapp"</span><span class="tsd-signature-symbol"> = &quot;fractapp&quot;</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/fractapp/fractapp/blob/a6c910d/src/utils/backup.ts#L15">src/utils/backup.ts:15</a></li>
						</ul>
					</aside>
					<div class="tsd-comment tsd-typography">
						<div class="lead">
							<p>default google drive folder</p>
						</div>
					</div>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-variable tsd-parent-kind-namespace">
					<a name="algorithm" class="tsd-anchor"></a>
					<h3><span class="tsd-flag ts-flagConst">Const</span> algorithm</h3>
					<div class="tsd-signature tsd-kind-icon">algorithm<span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">"aes-128-ctr"</span><span class="tsd-signature-symbol"> = &quot;aes-128-ctr&quot;</span></div>
					<aside class="tsd-sources">
						<ul>
							<li>Defined in <a href="https://github.com/fractapp/fractapp/blob/a6c910d/src/utils/backup.ts#L28">src/utils/backup.ts:28</a></li>
						</ul>
					</aside>
					<div class="tsd-comment tsd-typography">
						<div class="lead">
							<p>encryption algorithm</p>
						</div>
					</div>
				</section>
			</section>
			<section class="tsd-panel-group tsd-member-group ">
				<h2>Functions</h2>
				<section class="tsd-panel tsd-member tsd-kind-function tsd-parent-kind-namespace">
					<a name="backup-1" class="tsd-anchor"></a>
					<h3>backup</h3>
					<ul class="tsd-signatures tsd-kind-function tsd-parent-kind-namespace">
						<li class="tsd-signature tsd-kind-icon">backup<span class="tsd-signature-symbol">(</span>seed<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span>, password<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span>, type<span class="tsd-signature-symbol">: </span><a href="../enums/backup.backuptype.html" class="tsd-signature-type">BackupType</a><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-symbol">{ </span>fileName<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">; </span>isSuccess<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol"> }</span><span class="tsd-signature-symbol">&gt;</span></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/fractapp/fractapp/blob/a6c910d/src/utils/backup.ts#L43">src/utils/backup.ts:43</a></li>
								</ul>
							</aside>
							<div class="tsd-comment tsd-typography">
								<div class="lead">
									<p>backup file to phone drive or google</p>
								</div>
							</div>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>seed: <span class="tsd-signature-type">string</span></h5>
								</li>
								<li>
									<h5>password: <span class="tsd-signature-type">string</span></h5>
								</li>
								<li>
									<h5>type: <a href="../enums/backup.backuptype.html" class="tsd-signature-type">BackupType</a></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-symbol">{ </span>fileName<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">; </span>isSuccess<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol"> }</span><span class="tsd-signature-symbol">&gt;</span></h4>
						</li>
					</ul>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-function tsd-parent-kind-namespace">
					<a name="decrypt" class="tsd-anchor"></a>
					<h3>decrypt</h3>
					<ul class="tsd-signatures tsd-kind-function tsd-parent-kind-namespace">
						<li class="tsd-signature tsd-kind-icon">decrypt<span class="tsd-signature-symbol">(</span>file<span class="tsd-signature-symbol">: </span><a href="../classes/filebackup.html" class="tsd-signature-type">FileBackup</a>, password<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/fractapp/fractapp/blob/a6c910d/src/utils/backup.ts#L35">src/utils/backup.ts:35</a></li>
								</ul>
							</aside>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>file: <a href="../classes/filebackup.html" class="tsd-signature-type">FileBackup</a></h5>
								</li>
								<li>
									<h5>password: <span class="tsd-signature-type">string</span></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">string</span></h4>
						</li>
					</ul>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-function tsd-parent-kind-namespace">
					<a name="encrypt" class="tsd-anchor"></a>
					<h3>encrypt</h3>
					<ul class="tsd-signatures tsd-kind-function tsd-parent-kind-namespace">
						<li class="tsd-signature tsd-kind-icon">encrypt<span class="tsd-signature-symbol">(</span>seed<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span>, password<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><a href="../classes/filebackup.html" class="tsd-signature-type">FileBackup</a></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/fractapp/fractapp/blob/a6c910d/src/utils/backup.ts#L30">src/utils/backup.ts:30</a></li>
								</ul>
							</aside>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>seed: <span class="tsd-signature-type">string</span></h5>
								</li>
								<li>
									<h5>password: <span class="tsd-signature-type">string</span></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <a href="../classes/filebackup.html" class="tsd-signature-type">FileBackup</a></h4>
						</li>
					</ul>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-function tsd-parent-kind-namespace">
					<a name="getfile" class="tsd-anchor"></a>
					<h3>get<wbr>File</h3>
					<ul class="tsd-signatures tsd-kind-function tsd-parent-kind-namespace">
						<li class="tsd-signature tsd-kind-icon">get<wbr>File<span class="tsd-signature-symbol">(</span>filePath<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><a href="../classes/filebackup.html" class="tsd-signature-type">FileBackup</a><span class="tsd-signature-symbol">&gt;</span></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/fractapp/fractapp/blob/a6c910d/src/utils/backup.ts#L80">src/utils/backup.ts:80</a></li>
								</ul>
							</aside>
							<div class="tsd-comment tsd-typography">
								<div class="lead">
									<p>get the file on the device</p>
								</div>
							</div>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>filePath: <span class="tsd-signature-type">string</span></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><a href="../classes/filebackup.html" class="tsd-signature-type">FileBackup</a><span class="tsd-signature-symbol">&gt;</span></h4>
						</li>
					</ul>
				</section>
				<section class="tsd-panel tsd-member tsd-kind-function tsd-parent-kind-namespace">
					<a name="getseed" class="tsd-anchor"></a>
					<h3>get<wbr>Seed</h3>
					<ul class="tsd-signatures tsd-kind-function tsd-parent-kind-namespace">
						<li class="tsd-signature tsd-kind-icon">get<wbr>Seed<span class="tsd-signature-symbol">(</span>file<span class="tsd-signature-symbol">: </span><a href="../classes/filebackup.html" class="tsd-signature-type">FileBackup</a>, password<span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">&gt;</span></li>
					</ul>
					<ul class="tsd-descriptions">
						<li class="tsd-description">
							<aside class="tsd-sources">
								<ul>
									<li>Defined in <a href="https://github.com/fractapp/fractapp/blob/a6c910d/src/utils/backup.ts#L89">src/utils/backup.ts:89</a></li>
								</ul>
							</aside>
							<div class="tsd-comment tsd-typography">
								<div class="lead">
									<p>decrypt the file and get seed</p>
								</div>
							</div>
							<h4 class="tsd-parameters-title">Parameters</h4>
							<ul class="tsd-parameters">
								<li>
									<h5>file: <a href="../classes/filebackup.html" class="tsd-signature-type">FileBackup</a></h5>
								</li>
								<li>
									<h5>password: <span class="tsd-signature-type">string</span></h5>
								</li>
							</ul>
							<h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">&gt;</span></h4>
						</li>
					</ul>
				</section>
			</section>
		</div>
		<div class="col-4 col-menu menu-sticky-wrap menu-highlight">
			<nav class="tsd-navigation primary">
				<ul>
					<li class="globals  ">
						<a href="../globals.html"><em>Globals</em></a>
					</li>
					<li class=" tsd-kind-namespace">
						<a href="accountsstore.html">Accounts<wbr>Store</a>
					</li>
					<li class=" tsd-kind-namespace">
						<a href="authstore.html">Auth<wbr>Store</a>
					</li>
					<li class="current tsd-kind-namespace">
						<a href="backup.html">Backup</a>
					</li>
					<li class=" tsd-kind-namespace">
						<a href="db.html">DB</a>
					</li>
					<li class=" tsd-kind-namespace">
						<a href="date.html">Date</a>
					</li>
					<li class=" tsd-kind-namespace">
						<a href="dialogstore.html">Dialog<wbr>Store</a>
					</li>
					<li class=" tsd-kind-namespace">
						<a href="google.html">Google</a>
					</li>
					<li class=" tsd-kind-namespace">
						<a href="pricesstore.html">Prices<wbr>Store</a>
					</li>
					<li class=" tsd-kind-namespace">
						<a href="task.html">Task</a>
					</li>
				</ul>
			</nav>
			<nav class="tsd-navigation secondary menu-sticky">
				<ul class="before-current">
					<li class=" tsd-kind-enum tsd-parent-kind-namespace">
						<a href="../enums/backup.backuptype.html" class="tsd-kind-icon">Backup<wbr>Type</a>
					</li>
					<li class=" tsd-kind-variable tsd-parent-kind-namespace">
						<a href="backup.html#googledrivefolder" class="tsd-kind-icon">Google<wbr>Drive<wbr>Folder</a>
					</li>
					<li class=" tsd-kind-variable tsd-parent-kind-namespace">
						<a href="backup.html#algorithm" class="tsd-kind-icon">algorithm</a>
					</li>
					<li class=" tsd-kind-function tsd-parent-kind-namespace">
						<a href="backup.html#backup-1" class="tsd-kind-icon">backup</a>
					</li>
					<li class=" tsd-kind-function tsd-parent-kind-namespace">
						<a href="backup.html#decrypt" class="tsd-kind-icon">decrypt</a>
					</li>
					<li class=" tsd-kind-function tsd-parent-kind-namespace">
						<a href="backup.html#encrypt" class="tsd-kind-icon">encrypt</a>
					</li>
					<li class=" tsd-kind-function tsd-parent-kind-namespace">
						<a href="backup.html#getfile" class="tsd-kind-icon">get<wbr>File</a>
					</li>
					<li class=" tsd-kind-function tsd-parent-kind-namespace">
						<a href="backup.html#getseed" class="tsd-kind-icon">get<wbr>Seed</a>
					</li>
				</ul>
			</nav>
		</div>
	</div>
</div>
<footer class="with-border-bottom">
	<div class="container">
		<h2>Legend</h2>
		<div class="tsd-legend-group">
			<ul class="tsd-legend">
				<li class="tsd-kind-constructor tsd-parent-kind-class"><span class="tsd-kind-icon">Constructor</span></li>
				<li class="tsd-kind-property tsd-parent-kind-class"><span class="tsd-kind-icon">Property</span></li>
				<li class="tsd-kind-method tsd-parent-kind-class"><span class="tsd-kind-icon">Method</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-property tsd-parent-kind-class tsd-is-private"><span class="tsd-kind-icon">Private property</span></li>
			</ul>
			<ul class="tsd-legend">
				<li class="tsd-kind-method tsd-parent-kind-class tsd-is-static"><span class="tsd-kind-icon">Static method</span></li>
			</ul>
		</div>
	</div>
</footer>
<div class="container tsd-generator">
	<p>Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p>
</div>
<div class="overlay"></div>
<script src="../assets/js/main.js"></script>
</body>
</html>