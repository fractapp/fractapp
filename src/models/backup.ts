export class FileBackup {
    seed: string;
    algorithm: string;
    constructor(seed: string, algorithm: string) {
        this.seed = seed;
        this.algorithm = algorithm;
    }
}