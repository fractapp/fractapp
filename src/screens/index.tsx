export { default as ImportWallet } from './ImportWallet';
export { default as SaveSeed } from './SaveSeed';

export { default as SaveWallet } from './SaveWallet';
export { default as SettingWallet } from './SettingWallet';
export { default as Start } from './Start';

export { default as WalletDetails } from './WalletDetails';

export { default as WalletEncryption } from './WalletEncryption';
export { default as Wallets } from './Wallets';
export { ConfirmSaveSeed } from './ConfirmSaveSeed'