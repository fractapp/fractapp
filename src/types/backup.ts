/**
 * File backup model
 * @category Models
 */
export type FileBackup = {
  seed: string;
  algorithm: string;
};
