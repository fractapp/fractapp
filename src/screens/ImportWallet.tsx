import React, {useContext} from 'react';
import {StyleSheet, View, Text, Alert, PermissionsAndroid} from 'react-native';
import DocumentPicker from 'react-native-document-picker';
import {WhiteButton, Img} from 'components';
import backupUtil from 'utils/backup';
import {FileBackup} from 'models/backup';
import googleUtil from 'utils/google';
import Dialog from 'storage/Dialog';

export const ImportWallet = ({navigation}: {navigation: any}) => {
  const dialogContext = useContext(Dialog.Context);

  const openFilePicker = async () => {
    const statues = await PermissionsAndroid.requestMultiple([
      PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE,
      PermissionsAndroid.PERMISSIONS.READ_EXTERNAL_STORAGE,
    ]);
    let isGaranted = true;
    for (let key in statues) {
      const status = statues[key];
      if (status == 'granted') {
        continue;
      }
      if (status == 'never_ask_again') {
        dialogContext.dispatch(
          Dialog.open(
            'Open settings',
            'If you want to import a file then open the application settings and give it access to the storage.',
            () => dialogContext.dispatch(Dialog.close()),
          ),
        );
      }

      isGaranted = false;
    }

    if (!isGaranted) {
      return;
    }

    const res = await DocumentPicker.pick({
      type: [DocumentPicker.types.allFiles],
    });

    let file: FileBackup;
    try {
      file = await backupUtil.getFile(res.uri);
    } catch (err) {
      Alert.alert('Error', 'Invalid file');
      return;
    }

    navigation.navigate('WalletFileImport', {file: file});
  };

  const openFileGoogleDrivePicker = async () => {
    await googleUtil.signOut();
    await googleUtil.signIn();
    navigation.navigate('GoogleDrivePicker');
  };

  return (
    <View
      style={{
        flexDirection: 'column',
        flex: 1,
        alignItems: 'center',
        marginTop: 40,
      }}>
      <Text style={styles.title}>Import a wallet</Text>
      <View style={{width: '100%', alignItems: 'center', marginTop: 70}}>
        <WhiteButton
          text={'Enter seed'}
          height={50}
          img={Img.Key}
          width="90%"
          onPress={() => navigation.navigate('ImportSeed')}
        />
        <View style={{marginTop: 20, width: '90%'}}>
          <WhiteButton
            text={'From file'}
            img={Img.File}
            height={50}
            onPress={() => openFilePicker()}
          />
        </View>
        {
          <View style={{marginTop: 20, width: '90%'}}>
            <WhiteButton
              text={'Google drive'}
              img={Img.GoogleDrive}
              height={50}
              onPress={() => openFileGoogleDrivePicker()}
            />
          </View>
        }
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  title: {
    marginTop: 80,
    fontSize: 25,
    fontFamily: 'Roboto-Regular',
    color: '#2AB2E2',
  },
});
