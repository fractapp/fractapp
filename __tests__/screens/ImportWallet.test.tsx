import React from 'react';
import { PermissionsAndroid } from 'react-native'
import { render, fireEvent } from '@testing-library/react-native';
import { ImportWallet } from 'screens/ImportWallet';
import renderer from 'react-test-renderer';
import DialogStore from 'storage/Dialog'
import googleUtil from 'utils/google'
import DocumentPicker from 'react-native-document-picker';
import backupUtil from 'utils/backup';
import { FileBackup } from 'models/backup';

jest.mock('react-native', () => {
  const result = jest.requireActual('react-native')
  result.PermissionsAndroid.PERMISSIONS = {
    WRITE_EXTERNAL_STORAGE: "WRITE_EXTERNAL_STORAGE",
    READ_EXTERNAL_STORAGE: "READ_EXTERNAL_STORAGE"
  }
  result.PermissionsAndroid.requestMultiple = jest.fn()
  return result
})
jest.mock('utils/google', () => ({
  signIn: jest.fn(),
  signOut: jest.fn()
}))
jest.mock('utils/backup', () => ({
  getFile: jest.fn()
}))
jest.mock('react-native-document-picker', () => ({
  pick: jest.fn(),
  types: {
    allFiles: jest.fn()
  }
}))

it('Test positive', () => {
  const tree = renderer
    .create(
      <DialogStore.Context.Provider value={{
        dialog: {
          text: "",
          title: "",
          visible: false
        }
      }}>
        <ImportWallet navigation={null} />
      </DialogStore.Context.Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('Test click google drive', async () => {
  const mockFn = jest.fn()
  const component = render(
    <DialogStore.Context.Provider value={{
      dialog: {
        text: "",
        title: "",
        visible: false
      }
    }}>
      <ImportWallet navigation={{ navigate: mockFn }} />
    </DialogStore.Context.Provider>
  )

  await fireEvent.press(component.getByText('Google drive'));
  expect(googleUtil.signOut).toBeCalled()
  expect(googleUtil.signIn).toBeCalled()
  expect(mockFn).toBeCalledWith("GoogleDrivePicker")
});

it('Test click enter seed', () => {
  const mockFn = jest.fn()
  const component = render(
    <DialogStore.Context.Provider value={{
      dialog: {
        text: "",
        title: "",
        visible: false
      }
    }}>
      <ImportWallet navigation={{ navigate: mockFn }} />
    </DialogStore.Context.Provider>
  )

  fireEvent.press(component.getByText('Enter seed'));
  expect(mockFn).toBeCalledWith("ImportSeed")
});

it('Test click from file isGaranted=false', async () => {
  const mockNavFn = jest.fn()
  const mockDispatchFn = jest.fn()

  PermissionsAndroid.requestMultiple.mockReturnValueOnce({
    "WRITE_EXTERNAL_STORAGE": "denied",
    "READ_EXTERNAL_STORAGE": "denied"
  })

  const file = new FileBackup("seed", "algorithm")
  backupUtil.getFile.mockReturnValueOnce(file)
  const component = render(
    <DialogStore.Context.Provider value={{
      dialog: {
        text: "",
        title: "",
        visible: false
      },
      dispatch: mockDispatchFn
    }}>
      <ImportWallet navigation={{ navigate: mockNavFn }} />
    </DialogStore.Context.Provider>
  )

  await fireEvent.press(component.getByText('From file'));

  expect(PermissionsAndroid.requestMultiple).toBeCalledWith(["WRITE_EXTERNAL_STORAGE", "READ_EXTERNAL_STORAGE"])
  expect(DocumentPicker.pick).not.toBeCalled()
  expect(backupUtil.getFile).not.toBeCalled()
  expect(mockNavFn).not.toBeCalledWith("WalletFileImport", { file: file })
});

it('Test click from file isGaranted=false && open dialog', async () => {
  const mockNavFn = jest.fn()
  const mockDispatchFn = jest.fn()

  PermissionsAndroid.requestMultiple.mockReturnValueOnce({
    "WRITE_EXTERNAL_STORAGE": "never_ask_again",
    "READ_EXTERNAL_STORAGE": "never_ask_again"
  })

  const file = new FileBackup("seed", "algorithm")
  backupUtil.getFile.mockReturnValueOnce(file)
  const component = render(
    <DialogStore.Context.Provider value={{
      dialog: {
        text: "",
        title: "",
        visible: false
      },
      dispatch: mockDispatchFn
    }}>
      <ImportWallet navigation={{ navigate: mockNavFn }} />
    </DialogStore.Context.Provider>
  )

  await fireEvent.press(component.getByText('From file'));

  expect(PermissionsAndroid.requestMultiple).toBeCalledWith(["WRITE_EXTERNAL_STORAGE", "READ_EXTERNAL_STORAGE"])
  expect(mockDispatchFn).toBeCalled()
  expect(DocumentPicker.pick).not.toBeCalled()
  expect(backupUtil.getFile).not.toBeCalled()
  expect(mockNavFn).not.toBeCalledWith("WalletFileImport", { file: file })
});

it('Test click from file isGaranted=true', async () => {
  const mockNavFn = jest.fn()
  const mockDispatchFn = jest.fn()

  PermissionsAndroid.requestMultiple.mockReturnValueOnce({
    "WRITE_EXTERNAL_STORAGE": "granted",
    "READ_EXTERNAL_STORAGE": "granted"
  })

  const file = new FileBackup("seed", "algorithm")
  backupUtil.getFile.mockReturnValueOnce(file)
  DocumentPicker.pick.mockReturnValueOnce({ uri: "" })
  
  const component = render(
    <DialogStore.Context.Provider value={{
      dialog: {
        text: "",
        title: "",
        visible: false
      },
      dispatch: mockDispatchFn
    }}>
      <ImportWallet navigation={{ navigate: mockNavFn }} />
    </DialogStore.Context.Provider>
  )

  await fireEvent.press(component.getByText('From file'));

  expect(PermissionsAndroid.requestMultiple).toBeCalledWith(["WRITE_EXTERNAL_STORAGE", "READ_EXTERNAL_STORAGE"])
  expect(DocumentPicker.pick).toBeCalled()
  expect(backupUtil.getFile).toBeCalled()
  expect(mockNavFn).toBeCalledWith("WalletFileImport", { file: file })
});
