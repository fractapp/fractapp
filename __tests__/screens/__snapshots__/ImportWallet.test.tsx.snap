// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test positive 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "flexDirection": "column",
      "marginTop": 40,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#2AB2E2",
        "fontFamily": "Roboto-Regular",
        "fontSize": 25,
        "marginTop": 80,
      }
    }
  >
    Import a wallet
  </Text>
  <View
    style={
      Object {
        "alignItems": "center",
        "marginTop": 70,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderColor": "black",
          "borderRadius": 10,
          "borderWidth": 0.5,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "width": "90%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "#888888",
              "fontSize": 30,
            },
            Object {
              "left": 70,
              "position": "absolute",
            },
            Object {
              "fontFamily": "Material Design Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        󰷖
      </Text>
      <Text
        style={
          Object {
            "color": "black",
            "fontFamily": "Roboto-Regular",
            "fontSize": 15,
            "fontStyle": "normal",
            "fontWeight": "normal",
          }
        }
      >
        Enter seed
      </Text>
    </View>
    <View
      style={
        Object {
          "marginTop": 20,
          "width": "90%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 10,
            "borderWidth": 0.5,
            "height": 50,
            "justifyContent": "center",
            "opacity": 1,
            "width": "100%",
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#888888",
                "fontSize": 30,
              },
              Object {
                "left": 70,
                "position": "absolute",
              },
              Object {
                "fontFamily": "Material Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
        <Text
          style={
            Object {
              "color": "black",
              "fontFamily": "Roboto-Regular",
              "fontSize": 15,
              "fontStyle": "normal",
              "fontWeight": "normal",
            }
          }
        >
          From file
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "marginTop": 20,
          "width": "90%",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "borderColor": "black",
            "borderRadius": 10,
            "borderWidth": 0.5,
            "height": 50,
            "justifyContent": "center",
            "opacity": 1,
            "width": "100%",
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../assets/img/google-drive.png",
            }
          }
          style={
            Array [
              Object {
                "left": 70,
                "position": "absolute",
              },
              Object {
                "height": 30,
                "width": 30,
              },
            ]
          }
        />
        <Text
          style={
            Object {
              "color": "black",
              "fontFamily": "Roboto-Regular",
              "fontSize": 15,
              "fontStyle": "normal",
              "fontWeight": "normal",
            }
          }
        >
          Google drive
        </Text>
      </View>
    </View>
  </View>
</View>
`;
